test/regex.spec.js

```diff
@@ -0,0 +1,87 @@
+import assert from "assert";
+
+describe("RegEx", function () {
+  it("Should run exec() success", function () {
+    const regex = /foo*/g;
+    const str = "table football, foosball";
+    let match = regex.exec(str);
+    let index = [];
+    while (match) {
+      index.push(regex.lastIndex);
+      match = regex.exec(str);
+    }
+    assert.deepStrictEqual(index, [9, 19]);
+  });
+
+  it("Should run exec() fail", function () {
+    const regex = /foo*/g;
+    const str = "table dog, cat";
+    const match = regex.exec(str);
+    assert.deepStrictEqual(match, null);
+    assert.deepStrictEqual(regex.lastIndex, 0);
+  });
+
+  it("Should run test()", function () {
+    const str = "table football";
+    const regex = /foo*/;
+    const globalRegex = /foo*/g;
+
+    assert.strictEqual(regex.test(str), true);
+    assert.strictEqual(globalRegex.lastIndex, 0);
+    assert.strictEqual(globalRegex.test(str), true);
+    assert.strictEqual(globalRegex.lastIndex, 9);
+    assert.strictEqual(globalRegex.test(str), false);
```

Comment:
Just want to emphasize that the internal state of a regex object can be a cause for bugs. For this reason I just about always recommend creating regex objects inline. For example:

```js
// Not recommended
const emailRegex = /[\w\d_]+@[\w\d_]+.[\w\d]/
function isProbablyEmail(str) {
  return emailRegex.test(str)
}

// Recommended
function isProbablyEmail(str) {
  return /[\w\d_]+@[\w\d_]+.[\w\d]/.test(str)
}
```

Reply:
changed to inline format for wherever it can apply

Reply:
for this `globalRegex`, this remains as a variable to check the change of `lastIndex` each time after it runs `test()`